src/builtins/BuiltinAtomics.cpp

```diff
@@ -435,6 +436,179 @@ static Value builtinAtomicsXor(ExecutionState& state, Value thisValue, size_t ar
+    WL->m_waiterCount++;
+    WL->m_mutex.unlock();
+    if (t == std::numeric_limits<double>::infinity()) {
+        WL->m_waiter.wait(ul);
+    } else {
+        notified = WL->m_waiter.wait_for(ul, std::chrono::milliseconds((int64_t)t)) == std::cv_status::no_timeout;
+    }
+    WL->m_mutex.lock();
+    // 20. If notified is true, then
+    //     a. Assert: W is not on the list of waiters in WL.
+    // 21. Else,
+    //     a. Perform RemoveWaiter(WL, W).
+    if (!notified) {
+        WL->m_waiterCount--;
+    }
+    // 22. Perform LeaveCriticalSection(WL).
+    WL->m_mutex.unlock();
+    // 23. If notified is true, return the String "ok".
+    if (notified) {
+        return Value(state.context()->staticStrings().lazyOk().string());
+    }
+    // 24. Return the String "timed-out".
+    return Value(state.context()->staticStrings().lazyTimedOut().string());
+}
+
+static Value builtinAtomicsNotify(ExecutionState& state, Value thisValue, size_t argc, Value* argv, Optional<Object*> newTarget)
+{
+    // https://tc39.es/ecma262/#sec-atomics.notify
+    // Atomics.notify ( typedArray, index, count )
+    // 1. Let buffer be ? ValidateIntegerTypedArray(typedArray, true).
+    ArrayBuffer* buffer = validateIntegerTypedArray(state, argv[0], true);
+    TypedArrayObject* typedArray = argv[0].asObject()->asTypedArrayObject();
+    // 2. Let indexedPosition be ? ValidateAtomicAccess(typedArray, index).
+    size_t indexedPosition = validateAtomicAccess(state, typedArray, argv[1]);
+    double c;
+    // 3. If count is undefined, let c be +∞.
+    if (argv[2].isUndefined()) {
+        c = std::numeric_limits<double>::infinity();
+    } else {
+        // 4. Else,
+        // a. Let intCount be ? ToIntegerOrInfinity(count).
+        auto intCount = argv[2].toInteger(state);
+        // b. Let c be max(intCount, 0).
+        c = std::max(intCount, 0.0);
+    }
+    // 5. Let block be buffer.[[ArrayBufferData]].
+    void* blockAddress = reinterpret_cast<int32_t*>(buffer->data()) + indexedPosition;
+    // 6. Let arrayTypeName be typedArray.[[TypedArrayName]].
+    // 7. If IsSharedArrayBuffer(buffer) is false, return +0𝔽.
+    if (!buffer->isSharedArrayBufferObject()) {
+        return Value(0);
+    }
+    // 8. Let WL be GetWaiterList(block, indexedPosition).
+    Global::Waiter* WL = Global::waiter(blockAddress);
+    // 9. Let n be 0.
+    double n = 0;
+    // 10. Perform EnterCriticalSection(WL).
+    WL->m_mutex.lock();
+    double count = std::min((double)WL->m_waiterCount, c);
+    WL->m_waiterCount -= count;
+    // 11. Let S be RemoveWaiters(WL, c).
+    // 12. Repeat, while S is not an empty List,
+    //     a. Let W be the first agent in S.
+    //     b. Remove W from the front of S.
+    //     c. Perform NotifyWaiter(WL, W).
+    //     d. Set n to n + 1.
+    for (n = 0; n < count; n++) {
+        WL->m_mutex.unlock();
+        WL->m_waiter.notify_one();
+        WL->m_mutex.lock();
```

Comment:
Is it necessary to repeatedly lock and unlock between the notify operation here?

Reply:
unlock, lock is needed by spec. :(

Reply:
I see :)